### PR TITLE
[geojson] Fix nullable geometries & invalid properties generics

### PR DESCRIPTION
**Pull request by reilem**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://datatracker.ietf.org/doc/html/rfc7946#section-3.2 

This PR makes two main changes:
1. Make sure the properties object in a feature cannot be anything other than a JSON object (technically it can now still be an array but fixing this might be a bit overkill). There was a bug in the generics, the properties object was correctly assigned as default but not restricted with an `extends` so you could still just pass any random thing as the properties type. This has now been fixed. Quote from the RFC:
> A Feature object has a member with the name "properties".  The
      value of the properties member is an object (any JSON object or a
      JSON null value).

In the current version this would be "correctly typed". However, if you try to pass this into a GeoJSON validator it would not be allowed:
```typescript
const test: Feature<Geometry, number> = { type: "Feature", geometry: ..., properties: 567 }
```
2. Make sure that feature with null geometries can be assigned to the GeoJSON type. The default `G` type parameter in `Feature` and `FeatureCollection` was `Geometry` instead of `Geometry | null`. And since `GeoJSON` could also not take any generics this was not modifiable when using the `GeoJSON` type. I fixed both of these things by changing the default and by making GeoJSON generic so it can pass down the geometry type you want to use. Quote from the RFC:
> A Feature object has a member with the name "geometry".  The value
      of the geometry member SHALL be either a Geometry object as
      defined above or, in the case that the Feature is unlocated, a
      JSON null value.

With the way things were this was throwing a typing error, when in reality it should be allowed:
```typescript
function foo(geojson: GeoJSON) { ... };
foo({ type: "Feature", geometry: null, properties: null }); // TS2322: Type null is not assignable to type Geometry
```

NOTE: These changes could potentially be breaking if users have been reliant on the incorrect defaults & types. I'm not sure what you guys prefer, if we should keep it as-is so as not to annoy people, or if we should adhere to the RFC.

As a middle ground I would be willing to revert the property object and default geometry types change if this causes too much trouble for people. The main reason for this PR is to make GeoJSON generic so that we can use nullable geometries in GeoJSON.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/leaflet/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against a11e968dd0b22fb3327162b3188e3e1458536dbb
</p>

**Comment by brookjordan** [comment]
I worry that this could be quite a heavily breaking change.

While it does look like properties is supposed to only be an object, I know I've used it as a string in the past, and I doubt I'm the only one relying on this being “valid”. How to test my guess, however, I don't know.

I wonder if it would be safer for someone who needs this rigidness to wrap the type in a stricter one in their own codebase? I can't think of a way, off the top of my head, to opt in/out of this. But I fo think this would at least need to be sem-vered as a breaking change.

**Comment by reilem** [comment]
I have the same worry. Once I started fixing all the dependencies I understood the scope of this change. If others agree I will probably revert the properties change. The type is correct by default, if users wish to break this then it's at their own risk.

**Review by danvk — COMMENTED** [review]
Is there any reason these changes need to be bundled together? If they were split, we could discuss them independently.

The `type GeoJSON` change seems fine to me, though in all honesty I've never used this type. I usually know whether I have a Feature, FeatureCollection, or Geometry.

The `GeoJsonProperties` change doesn't feel like a good tradeoff between breaking existing code and catching real bugs. But if anyone winds up making a new, more precise GeoJSON types library, it should certainly include this change from the start! (See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70029#issuecomment-2231830232).

File: types/geojson/index.d.ts [hunk 1]
```diff
@@ -60,7 +60,10 @@ export interface GeoJsonObject {
 /**
  * Union of GeoJSON objects.
  */
-export type GeoJSON = Geometry | Feature | FeatureCollection;
+export type GeoJSON<G extends Geometry | null = Geometry | null, P extends GeoJsonProperties = GeoJsonProperties> =
+    | Geometry
```
Comment (danvk) [inline types/geojson/index.d.ts:64]: Would it make more sense for this to be `G`? Or at least `Exclude<G, null>`?

This feels a little funny to allow:

```ts
const geojson: GeoJSON<Point> = lineString;
```

Comment (reilem) [reply types/geojson/index.d.ts:64]: You are entirely correct! I will make this change

**Comment by reilem** [comment]
I will omit the properties change for now. Certainly not as critical as the other change.

**Comment by danvk** [comment]
Sorry for the slow response here. I tried plugging this change into one of my projects and got something like 15-20 errors. These were all due to changing the default geometry type of `Feature` and `FeatureCollection` to include `null`. My project does not use `null` geometries, so these were all false positives.

The change to make `GeoJSON` generic seems 100% non-controversial, and we should definitely do that.

I'm not convinced the change to include `null` geometries by default is worth the disruption it will cause, though. Many (most?) projects do not use `null` geometries, and this change will be disruptive to them. Projects that do want to consider `null` geometries can already opt in by using `Feature<Geometry | null>`.

At the very least this would be a breaking change, which would require bumping the major version. But then we get into the business of the weird version number of this package.

So I am 👍 to the `GeoJSON` change, but 👎 to the `Feature` and `FeatureCollection` changes.

**Comment by underbluewaters** [comment]
I concur with everything @danvk said. I'm not 100% opposed to adding null geometry to the type if there is a use-case (it is in the spec after all), just in practice I never see data like that. I'm not sure it's worth the extra work to handle this hypothetical.

**Comment by reilem** [comment]
Thanks for the comments. I agree with the suggestions and will make the change.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/geojson/index.d.ts
1. types/geojson/geojson-tests.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e5da3e329025a13357d40fdba135dc438f1886ff
</p>

**Review by danvk — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]

**Comment by reilem** [comment]
Ready to merge

**Pull request merged**